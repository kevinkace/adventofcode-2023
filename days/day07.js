import { eol } from "../consts";

const cardValuesDesc = [ "A", "K", "Q", "J", "T", "9", "8", "7", "6", "5", "4", "3", "2" ];

const faceCardValues = {
    A: 14,
    K: 13,
    Q: 12,
    J: 11,
    T: 10
};

const handValues = [
    "11111",
    "2111",
    "221",
    "311",
    "32",
    "41",
    "5"
];

export class Games {
    constructor(str) {
        this.str = str;

        this.rounds = this.parse();
    }

    parse() {
        return this.str
            .split(eol)
            .map(line =>{
                const [ cards, bet ] = line.split(" ");

                const cardsInOrder = cards.split("");

                const grouped = cardsInOrder
                    .reduce((acc, card) => {
                        acc[card] = acc[card] ? acc[card] + 1 : 1;

                        return acc;
                    }, {});

                const value = handValues.indexOf(Object.values(grouped).sort((a, b) => b - a).join(""));

                return { cards, grouped, bet : Number(bet), value, cardsInOrder };
            });
    }

    toString() {
        return this.rounds
            .reduce((acc, line) => {
                const row = [ line.cards, line.bet ].join(" ");

                return [ acc, row, eol ].join("");
            }, ""
            )
            .trim();
    }

    getSortedHands() {
        return this.rounds.slice().sort(Games.sortHand);
    }

    getSortedBets() {
        return this.getSortedHands().map(hand => hand.bet);
    }

    getTotalWinnings() {
        return this.getSortedHands().reduce((acc, hand, idx) => {
            return acc + (hand.bet * (idx + 1));
        }, 0);
    }

    static sortHand(hand1, hand2) {
        if (hand1.value > hand2.value) {
            return 1;
        } if (hand1.value < hand2.value) {
            return -1;
        }

        for (let idx = 0; idx < hand1.cardsInOrder.length; idx++) {
            const hand1Card = hand1.cardsInOrder[idx];
            const hand2Card = hand2.cardsInOrder[idx];

            if (hand1Card === hand2Card) {
                continue;
            }

            const card1Value = faceCardValues[hand1Card] || Number(hand1Card);
            const card2Value = faceCardValues[hand2Card] || Number(hand2Card);

            return card1Value > card2Value ? 1 : -1;
        }
    }
}
