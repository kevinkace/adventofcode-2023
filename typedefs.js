/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

export default {};
