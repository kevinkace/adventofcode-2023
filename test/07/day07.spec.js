import { readFile }      from "node:fs/promises";
import { fileURLToPath } from "node:url";
import path              from "node:path";

import { expect, test, describe } from "vitest";

import { Games } from "../../days/day07.js";

import typedefs from "../../typedefs.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const sample = await readFile(path.join(__dirname, "sample.txt"), "utf8");
const input  = await readFile(path.join(__dirname, "input.txt"), "utf8");


describe("day 07", () => {
    test("can parse input", () => {
        const sampleGame = new Games(sample);

        expect(sampleGame.toString()).toBe(sample);
    });

    test("can sort games", () => {
        const sampleGame = new Games(sample);

        expect(sampleGame.getSortedHands()).toMatchSnapshot();;
    });

    test("can sort bets", () => {
        const sampleGame = new Games(sample);

        expect(sampleGame.getSortedBets()).toMatchSnapshot();;
    });

    test("can get total winnings", () => {
        const sampleGame = new Games(sample);

        expect(sampleGame.getTotalWinnings()).toBe(6440);
    });

    test("answer p1", () => {
        const game = new Games(input);

        expect(game.getTotalWinnings()).toMatchSnapshot();
    });
});
