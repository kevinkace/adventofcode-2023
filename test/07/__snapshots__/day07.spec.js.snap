// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 07 > answer p1 1`] = `250120186`;

exports[`day 07 > can sort bets 1`] = `
[
  765,
  220,
  28,
  684,
  483,
]
`;

exports[`day 07 > can sort games 1`] = `
[
  {
    "bet": 765,
    "cards": "32T3K",
    "cardsInOrder": [
      "3",
      "2",
      "T",
      "3",
      "K",
    ],
    "grouped": {
      "2": 1,
      "3": 2,
      "K": 1,
      "T": 1,
    },
    "value": 1,
  },
  {
    "bet": 220,
    "cards": "KTJJT",
    "cardsInOrder": [
      "K",
      "T",
      "J",
      "J",
      "T",
    ],
    "grouped": {
      "J": 2,
      "K": 1,
      "T": 2,
    },
    "value": 2,
  },
  {
    "bet": 28,
    "cards": "KK677",
    "cardsInOrder": [
      "K",
      "K",
      "6",
      "7",
      "7",
    ],
    "grouped": {
      "6": 1,
      "7": 2,
      "K": 2,
    },
    "value": 2,
  },
  {
    "bet": 684,
    "cards": "T55J5",
    "cardsInOrder": [
      "T",
      "5",
      "5",
      "J",
      "5",
    ],
    "grouped": {
      "5": 3,
      "J": 1,
      "T": 1,
    },
    "value": 3,
  },
  {
    "bet": 483,
    "cards": "QQQJA",
    "cardsInOrder": [
      "Q",
      "Q",
      "Q",
      "J",
      "A",
    ],
    "grouped": {
      "A": 1,
      "J": 1,
      "Q": 3,
    },
    "value": 3,
  },
]
`;
